Add thread pool wrapper

lethosor opened the pull request:
Per discussion in #51, we might want a way to perform tasks that involve multiple API requests in parallel. #51's approach didn't make that easier, so this is a very small PR that makes a ThreadPoolExecutor available (across Python versions) and allows things like this (pseudocode):

```python
api = API()

def run_job(data):
  data_id = api.upload_data(data)
  job_id = api.run_job(data_id)
  return job_id

job_ids = list(api.thread_map(run_job, data_files))
```

By default, at most 5 calls to `run_job` would be active at a time, but this is adjustable.

Alternatively, `thread_pool().map()` could be abstracted away - [Executor](https://docs.python.org/dev/library/concurrent.futures.html#concurrent.futures.Executor) doesn't provide much else that would be useful in this case.

Commit:
Add thread pool wrapper

MikeJeffers commented:
> Alternatively, thread_pool().map() could be abstracted away - Executor doesn't provide much else that would be useful in this case.

Yes - think of how you would have to tell a client user how to use this library.
Should it be the snippet above? Or a new arg on a function signature `start_job(data_ids, multithread=true)`, or a new function that mirrors the singular operations: `start_job(data_id..) -> start_jobs(data_ids ...)`

Commit:
Replace thread_pool() with thread_map() wrapper and document

lethosor commented:
Example usage as of https://github.com/voxel51/api-py/pull/54/commits/784de740cce2cc8a657561992ea3275e7192905f:
```python
list(api.thread_map(api.upload_data, filenames))
```
or
```python
for data in api.thread_map(api.upload_data, filenames):
  print("Uploaded", data["id"])
```

Generators have some advantages (e.g. the second snippet won't wait for *all* upload_data calls to return before it starts printing) but can also be confusing (if you don't read from the generator, it won't call the callback at all). Thoughts?

Another option is to add convenience wrappers for some common cases (e.g. `upload_jobs`). Keeping `thread_map()` around is convenient for cases when users want to do something more complicated that involves multiple API calls (but I'd be up for a better name if you have ideas).

Commit:
Make thread_map return a list by default, add a generator arg

lethosor commented:
Added an option to return a generator instead of a list - returning a list is the default behavior now, as it's easier to deal with, and most use cases probably won't _need_ generators.

Updated usage as of https://github.com/voxel51/api-py/pull/54/commits/7c1dccafea2e8e51cd81278e54d7eba2dedc34f4:
```python
api.thread_map(api.upload_data, filenames)
```
or (using `generator` for efficiency - not strictly required):
```python
for data in api.thread_map(api.upload_data, filenames, generator=True):
  print("Uploaded", data["id"])
```

Commit:
Make thread_map a static method

Commit:
Note about max_workers=None

MikeJeffers reviewed (approved):
I'm ok with these changes with the argument it is adding a feature supported naturally in the js api.
Next: docs!

brimoor committed:
always returning a list, and using context manager

Commit:
Add note on thread_map to readme

Adapted from api-js

lethosor commented:
Adapted docs from the [JS client lib docs](https://github.com/voxel51/api-js#improving-request-efficiency)

brimoor committed:
using separate function to evoke imagination of the reader

brimoor commented:
why is everyone afraid to merge this? I'm pulling the trigger
